### PR TITLE
Add Delitoon websites (Delitoon/X/DE/Bomtoon/Bontoon/LezhinES)

### PR DESCRIPTION
**Pull request by MikeZeDev**
Most files are empty shells that just extends Delitoon :D

### CODE REVIEW

File: web/src/engine/websites/LezhinES.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { Tags } from '../Tags';
+import icon from './LezhinES.webp';
+import Delitoon from './Delitoon';
+export default class extends Delitoon {
```
Comment (ronny1982) [inline web/src/engine/websites/LezhinES.ts:4]: LezhinES is based on Delitoon and not affiliated with LezhinEN, LezhinKO and LezhinJA?

File: web/src/engine/websites/Delitoon.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        expiredAt: number
+    },
+    refreshToken: {
+        token: string
+    }
+}
+
+@Common.ImageAjax()
+export default class extends DecoratableMangaScraper {
+    protected BalconyID: string = undefined;
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:57]: Make fields private readonly and assign in constructor

File: web/src/engine/websites/Delitoon.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        return icon;
+    }
+
+    public override ValidateMangaURL(url: string): boolean {
+        return new RegExp(`^${this.URI.origin}/detail/[^/]+$`).test(url);
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const mangaid = new URL(url).href.match(/\/detail\/([^/]+)/)[1];
+        const req = new URL(`/api/balcony-api-v2/contents/${mangaid}`, this.URI);
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:77]: Variable name is misleading, i would expect a Request but it is an URL

File: web/src/engine/websites/Delitoon.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                const url = new URL('/api/balcony/auth/refresh', this.URI);
+                const body = JSON.stringify({
+                    accessToken: this.Token.accessToken.token,
+                    refreshToken: this.Token.refreshToken.token
+                });
+                const { result } = await FetchJSON<APIRefreshToken>(this.CreatePostRequest(url, body));
+                this.Token = result;
+            }
+        } catch (error) {
+            this.Token = undefined;
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:171]: Shouldn't the exception be forwarded to be handled by the caller?
All errors such as network error, out of memory, ... are just silently ignored

File: web/src/engine/websites/Delitoon.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        };
+        if (this.Token) headers['authorization'] = ' Bearer ' + this.Token.accessToken.token;
+        return new Request(url, {
+            method: 'POST',
+            headers: headers,
+            body: body,
+        });
+    }
+
+    protected async GetToken() {
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:149]: Method is called `GetToken` but is void.
Use a more appropriate name for the method.

File: web/src/engine/websites/Delitoon.ts [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            //if token is undefined => try to fetch one
+            if (!this.Token) {
+                const url = new URL('/api/auth/session', this.URI);
+                const { user } = await FetchJSON<APIUser>(this.CreateRequest(url));
+                this.Token = user;
+                return;
+            }
+
+            //if token is defined, check for expiration  and refresh if needed
+            if (this.Token && Date.now() > this.Token.accessToken.expiredAt) { //24h expiration
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:161]: The check for `this.Token` is redundant as it is guarded by the previous condition which checks for `!this.Token`.

File: web/src/engine/websites/Delitoon.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            contentsThumbnailType: 'MAIN'
+        }).toString();
+        const { data } = await FetchJSON<APIResult<APIManga[]>>(this.CreateRequest(url));
+        return data.map(element => new Manga(this, provider, element.alias, element.title.trim()));
+    }
+
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
+        await this.GetToken();
+        const url = new URL(`/api/balcony-api-v2/contents/${manga.Identifier}`, this.URI);
+        url.searchParams.set('isNotLoginAdult', 'true');
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:98]: Does this exclude all adult content?
How is adult content accessible.

File: web/src/engine/websites/Delitoon.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            let title = element.title.trim();
+            title += element.subTitle ? ' : ' + element.subTitle.trim() : '';
+            return new Chapter(this, manga, element.alias, title);
+        });
+    }
+
+    public override async FetchPages(chapter: Chapter): Promise<Page[]> {
+        await this.GetToken();
+        const url = new URL(`/api/balcony-api-v2/contents/${chapter.Parent.Identifier}/${chapter.Identifier}`, this.URI);
+        url.searchParams.set('isNotLoginAdult', 'true');
```
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:110]: Is this parameter required when fetching pages?

File: web/src/engine/websites/Delitoon.ts (hunk 6 above)
Comment (ronny1982) [inline web/src/engine/websites/Delitoon.ts:161]: Is this check timezone safe, or is the expiration from the server's timezone?

**Review by ronny1982 — APPROVED** [review]

File: web/src/engine/websites/LezhinES.ts (hunk 1 above)
Comment (MikeZeDev) [reply web/src/engine/websites/LezhinES.ts:4]: Exactly. Weird isnt it?

File: web/src/engine/websites/Delitoon.ts (hunk 7 above)
Comment (MikeZeDev) [reply web/src/engine/websites/Delitoon.ts:98]: No. From what i see, its a variable to get adult content without being logged.

File: web/src/engine/websites/Delitoon.ts (hunk 8 above)
Comment (MikeZeDev) [reply web/src/engine/websites/Delitoon.ts:110]: image

File: web/src/engine/websites/Delitoon.ts (hunk 6 above)
Comment (MikeZeDev) [reply web/src/engine/websites/Delitoon.ts:161]: expiredAt is UTC timestamp. Date.now is the same iirc.

File: web/src/engine/websites/Delitoon.ts (hunk 8 above)
Comment (MikeZeDev) [reply web/src/engine/websites/Delitoon.ts:110]: It shows that /viewer/ is now a valid endpoint for, perhaps all websites. Gotta correct that (old still working)

**Pull request merged**